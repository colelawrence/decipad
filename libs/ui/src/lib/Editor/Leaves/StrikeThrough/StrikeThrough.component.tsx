import { Text } from '@chakra-ui/react';

import { RenderLeafProps } from 'slate-react';

export const StrikeThrough = ({
  attributes,
  children,
}: RenderLeafProps): JSX.Element => {
  return (
    <Text as="span" textDecoration="line-through" {...attributes}>
      {children}
    </Text>
  );
};
