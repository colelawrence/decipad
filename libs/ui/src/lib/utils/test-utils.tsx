import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';
import { theme } from '../theme';

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      {children}
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
