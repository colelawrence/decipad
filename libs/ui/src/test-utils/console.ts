export const mockConsoleWarn = () => {
  const originalConsoleWarn = console.warn;
  const mockConsoleWarn: jest.MockedFunction<typeof console.warn> = jest.fn();

  beforeEach(() => {
    console.warn = jest.fn();
    mockConsoleWarn.mockClear();
  });
  afterEach(() => {
    console.warn = originalConsoleWarn;
  });

  return mockConsoleWarn;
};
