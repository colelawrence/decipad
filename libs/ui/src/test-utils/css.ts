import cssVarsPonyfill from 'css-vars-ponyfill';

export const findParentWithStyle = <P extends keyof CSSStyleDeclaration>(
  element: Element | null,
  propertyName: P
):
  | (Pick<CSSStyleDeclaration, P> & {
      element: Element;
      styles: CSSStyleDeclaration;
    })
  | null => {
  if (element === null) {
    return null;
  }
  const styles = getComputedStyle(element);
  if (!styles[propertyName]) {
    return findParentWithStyle(element.parentElement, propertyName);
  }
  return {
    [propertyName]: styles[propertyName],
    element,
    styles,
    // cannot type dynamic property key based on type parameter
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any;
};

export const applyCssVars = (): Promise<() => void> => {
  return new Promise((resolve, reject) => {
    const cleanup = () =>
      [...document.head.querySelectorAll('[data-cssvars="out"]')].forEach(
        (element) => element.remove()
      );
    cssVarsPonyfill({
      onError: reject,
      onComplete: () => resolve(cleanup),
    });
  });
};

export const applyPrefersColorScheme = async (
  colorScheme: 'dark' | 'light'
): Promise<() => void> => {
  // JSDOM ignores CSSOM modifications, so build a CSS string and create and append a style element instead
  const styleText = [...document.styleSheets]
    .flatMap((styleSheet) => [...styleSheet.cssRules])
    .flatMap((rule) => {
      if (rule.type === CSSRule.MEDIA_RULE) {
        const mediaRule = rule as CSSMediaRule;
        for (
          let mediaIndex = 0;
          mediaIndex < mediaRule.media.length;
          mediaIndex += 1
        ) {
          if (
            mediaRule.media[mediaIndex].replace(/\s/g, '') ===
            `(prefers-color-scheme:${colorScheme})`
          ) {
            return [...mediaRule.cssRules];
          }
        }
      }
      return [];
    })
    .flatMap((rule) => rule.cssText)
    .join(' ');

  const styleElement = document.createElement('style');
  styleElement.innerHTML = styleText;
  document.head.appendChild(styleElement);

  const cleanupCssVars = await applyCssVars();

  return () => {
    cleanupCssVars();
    styleElement.remove();
  };
};
