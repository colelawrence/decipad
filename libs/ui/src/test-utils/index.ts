export * from './menu';
export * from './language';
