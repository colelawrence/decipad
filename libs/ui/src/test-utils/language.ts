import {
  Result,
  runCode as run,
  SerializedTypeKind,
  serializeResult,
} from '@decipad/computer';

export async function runCode<T extends SerializedTypeKind>(
  code: string
): Promise<Result.Result<T>> {
  const result = await run(code);
  return serializeResult<T>(result.type, result.value);
}
