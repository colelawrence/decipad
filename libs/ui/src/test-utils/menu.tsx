import { FC } from 'react';
import * as RadixDropdown from '@radix-ui/react-dropdown-menu';

export const MenuWrapper: FC<React.PropsWithChildren<unknown>> = ({
  children,
}) => (
  <RadixDropdown.Root open>
    <RadixDropdown.Trigger css={{ height: 0 }}>&nbsp;</RadixDropdown.Trigger>
    <RadixDropdown.Content>{children}</RadixDropdown.Content>
  </RadixDropdown.Root>
);
