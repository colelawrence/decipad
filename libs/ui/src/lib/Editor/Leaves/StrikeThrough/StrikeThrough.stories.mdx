import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';
import { StrikeThrough } from './StrikeThrough.component';

<Meta
  title='Editor/Leaves/StrikeThrough'
  parameters={{
    docs: {
      source: {
        code: `<StrikeThrough
  attributes={{ 'data-slate-leaf': true }}
  leaf={{ text: 'StrikeThrough text' }}
  text={{ text: 'StrikeThrough text' }}
>
  StrikeThrough text
</StrikeThrough>`,
      },
    },
  }}
  argTypes={{
    children: {
      defaultValue: 'Hello World',
      description: 'The text content of the component',
      type: 'string',
    },
  }}
  component={StrikeThrough}
/>

# StrikeThrough

Cross out anything you regret with this incredible leaf!

export const Template = (args) => <StrikeThrough {...args} />;

<Canvas>
  <Story name='StrikeThrough'>{Template.bind({})}</Story>
</Canvas>

<Props story='StrikeThrough' />

