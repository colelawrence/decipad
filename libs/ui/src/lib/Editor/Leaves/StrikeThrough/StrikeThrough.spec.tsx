import React from 'react';
import { render } from 'test-utils';
import { StrikeThrough } from './StrikeThrough.component';

describe('StrikeThrough Leaf', () => {
  it('renders', () => {
    const { container } = render(
      <StrikeThrough
        attributes={{ 'data-slate-leaf': true }}
        leaf={{ text: 'StrikeThrough text' }}
        text={{ text: 'StrikeThrough text' }}
      >
        StrikeThrough text
      </StrikeThrough>
    );

    expect(container).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <StrikeThrough
        attributes={{ 'data-slate-leaf': true }}
        leaf={{ text: 'StrikeThrough text' }}
        text={{ text: 'StrikeThrough text' }}
      >
        StrikeThrough text
      </StrikeThrough>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the string', () => {
    const { getByText } = render(
      <StrikeThrough
        attributes={{ 'data-slate-leaf': true }}
        leaf={{ text: 'StrikeThrough text' }}
        text={{ text: 'StrikeThrough text' }}
      >
        StrikeThrough text
      </StrikeThrough>
    );

    getByText('StrikeThrough text');
  });
});
