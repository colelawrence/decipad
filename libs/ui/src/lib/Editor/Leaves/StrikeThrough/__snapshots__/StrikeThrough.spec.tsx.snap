// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrikeThrough Leaf matches snapshot 1`] = `
.emotion-0 {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

<div>
  <span
    class="chakra-text emotion-0"
    data-slate-leaf="true"
  >
    StrikeThrough text
  </span>
</div>
`;
